x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java

```diff
@@ -82,8 +89,23 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
 
     private void handleException(String actionType, RestRequest request, RestChannel channel, Exception e) {
         logger.debug(new ParameterizedMessage("{} failed for REST request [{}]", actionType, request.uri()), e);
+        final RestStatus restStatus = ExceptionsHelper.status(e);
         try {
-            channel.sendResponse(new BytesRestResponse(channel, e));
+            channel.sendResponse(new BytesRestResponse(channel, restStatus, e) {
+
+                @Override
+                protected ToXContent.Params paramsFromRequest(RestRequest restRequest) {
+                    ToXContent.Params params = restRequest;
+                    if (params.paramAsBoolean("error_trace", !REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT)
+                            && restStatus != RestStatus.UNAUTHORIZED) {
+                        params = new ToXContent.DelegatingMapParams(singletonMap(REST_EXCEPTION_SKIP_STACK_TRACE, "false"), params);
+                    } else {
+                        params = new ToXContent.DelegatingMapParams(singletonMap(REST_EXCEPTION_SKIP_STACK_TRACE, "true"), params);
+                    }
+                    return params;
+                }
```

Comment:
I wonder whether this could just be pulled up into `BytesRestResponse` itself. I think the intention is probably to not mix security related code with generic Rest code. But the duplication worries me and it is hard to be aware of the additional logic if you just read `BytesRestResponse`.

If pulling this up does not sound right to you, alternatively, we could add an additional method to `BytesRestResponse`, e.g. `allowsStackTrace()`, which always returns `true` and is overridden here to check the restStatus, i.e.:

For `BytesRestResponse`:
```java
private ToXContent.Params paramsFromRequest(RestRequest restRequest) {
    ToXContent.Params params = restRequest;
    if (params.paramAsBoolean("error_trace", !REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT)
        && allowsStackTrace()) {
                 ...
    }
            ...
}

protected boolean allowsStackTrace() { return true; }
```

And overridden as:
```java
channel.sendResponse(new BytesRestResponse(channel, e) {
    @Override
    protected boolean allowsStackTrace() { return status() != RestStatus.UNAUTHORIZED; }
})
```

I personally feel the additional method makes it easier to be aware of the possible extra check by just reading the `BytesRestResponse` class.

Reply:
How about an additional ctor in BytesRestResponse that takes a boolean that in principal overrides `REST_EXCEPTION_SKIP_STACK_TRACE`, `REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT` or `error_trace` so that you can simply call
```
final RestStatus restStatus = ExceptionsHelper.status(e);
final boolean hideStackTrace = restStatus == RestStatus.UNAUTHORIZED;
try {
    channel.sendResponse(new BytesRestResponse(channel, hideStackTrace, e));
```

Reply:
The `allowsStackTrace()` alternative sounds good, thanks!